src/GafferScene/MeshToPoints.cpp

```diff
@@ -126,25 +151,94 @@ IECore::ConstObjectPtr MeshToPoints::computeProcessedObject( const ScenePath &pa
 		return inputObject;
 	}
 
-	IECore::PointsPrimitivePtr result = new PointsPrimitive( mesh->variableSize( PrimitiveVariable::Vertex ) );
-	for( PrimitiveVariableMap::const_iterator it = mesh->variables.begin(), eIt = mesh->variables.end(); it != eIt; ++it )
+	IECore::PointsPrimitivePtr result;
+
+	if( modePlug()->getValue() == "polygon" )
 	{
-		PrimitiveVariable::Interpolation interpolation = it->second.interpolation;
-		switch( interpolation )
+
+		float rotation = IECore::degreesToRadians(rotationPlug()->getValue());
+		size_t numFaces = mesh->numFaces();
+		IECore::V3fVectorDataPtr positions = new IECore::V3fVectorData();
+		IECore::V3fVectorData::ValueType &positionVector = positions->writable();
+
+		positionVector.resize( numFaces );
+
+		const IECore::V3fVectorData *meshPositions = mesh->variableData<IECore::V3fVectorData>( "P" );
+
+		QuatfVectorDataPtr orientationData = new QuatfVectorData();
+		orientationData->writable().resize( numFaces );
+
+		IntVectorDataPtr idData = new IntVectorData();
+		idData->writable().resize(numFaces);
+
+		size_t vertex = 0;
+		for( size_t f = 0; f < numFaces; ++f )
 		{
-			case PrimitiveVariable::Uniform :
-			case PrimitiveVariable::FaceVarying :
-				// Skip these, since they make no sense
-				// on points.
-				continue;
-			case PrimitiveVariable::Varying :
-				interpolation = PrimitiveVariable::Vertex;
-				break;
-			default :
-				break;
+			int positionIndex = mesh->vertexIds()->readable()[vertex];
+
+			Imath::V3f p0 = meshPositions->readable()[positionIndex + 0];
+			Imath::V3f p1 = meshPositions->readable()[positionIndex + 1];
+			Imath::V3f p2 = meshPositions->readable()[positionIndex + 2];
+
+			positionVector[f] = p0;
+
+			Imath::V3f d02 = p2 - p0;
+			Imath::V3f d01 = p1 - p0;
+
+			Imath::V3f n = d02.cross( d01 );
+			n.normalize();
+			Imath::V3f t = d01;
+			t.normalize();
+
+			Imath::V3f b = n.cross( t );
+			b.normalize();
+
+			//@formatter:off
+			Imath::M44f matrix(
+				t.x, t.y, t.z, 0.0f,
+				b.x, b.y, b.z, 0.0f,
+				n.x, n.y, n.z, 0.0f,
+				0.0f, 0.0f, 0.0f, 1.0f
+			);
+			//@formatter:on
+
+			Imath::M44f preRotation;
+			preRotation.setAxisAngle(Imath::V3f(0.0f, 0.0f, 1.0f), rotation);
+
+			matrix = preRotation * matrix;
+			Imath::Quatf quat = Imath::extractQuat( matrix );
+
+			orientationData->writable()[f] = quat;
```

Comment:
It's definitely worth hoisting calls to `writable()` and `readable()` outside of any loop, because they do have measurable overhead. I tend to use a convention of `Ptr fooData` and `vector<T> &foo` to distinguish between the owning pointer and the reference.